server/router/src/main/java/org/infinispan/server/router/handlers/SniRouteHandler.java

```diff
@@ -0,0 +1,59 @@
+package org.infinispan.server.router.handlers;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInitializer;
+import io.netty.handler.ssl.SniHandler;
+import io.netty.handler.ssl.SslContext;
+import io.netty.util.DomainNameMapping;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.server.router.RoutingTable;
+import org.infinispan.server.router.logging.RouterLogger;
+import org.infinispan.server.router.routes.impl.NettyHandlerRouteDestination;
+import org.infinispan.server.router.routes.Route;
+import org.infinispan.server.router.routes.SniRouteSource;
+
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.Optional;
+
+public class SniRouteHandler extends SniHandler {
+
+    private static final RouterLogger logger = LogFactory.getLog(MethodHandles.lookup().lookupClass(), RouterLogger.class);
+
+    private final RoutingTable routingTable;
+
+    public SniRouteHandler(DomainNameMapping<? extends SslContext> mapping, RoutingTable routingTable) {
+        super(mapping);
+        this.routingTable = routingTable;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        super.decode(ctx, in, out);
+
+        if(isHandShaked()) {
```

Comment:
No, the `SniHandler`'s logic will handle that. We just silently hook into the process if everything is all right.